boa/src/builtins/json/mod.rs

```diff
@@ -33,19 +33,67 @@ mod tests;
 /// [spec]: https://tc39.es/ecma262/#sec-json.parse
 /// [mdn]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/parse
 // TODO: implement optional revever argument.
-pub fn parse(_: &mut Value, args: &[Value], _: &mut Interpreter) -> ResultValue {
+pub fn parse(_: &mut Value, args: &[Value], interpreter: &mut Interpreter) -> ResultValue {
     match serde_json::from_str::<JSONValue>(
         &args
             .get(0)
             .expect("cannot get argument for JSON.parse")
             .clone()
             .to_string(),
     ) {
-        Ok(json) => Ok(Value::from(json)),
+        Ok(json) => {
+            let j = Value::from(json);
+            if args.len() > 1 {
+                match args.get(1) {
+                    Some(callback) => {
+                        if callback.is_function() {
+                            let mut holder = Value::new_object(None);
+                            holder.set_field(Value::from(""), j.clone());
+                            walk(callback, interpreter, &mut holder, Value::from(""))
+                        } else {
+                            Ok(j)
+                        }
+                    }
+                    _ => Ok(j),
+                }
+            } else {
+                Ok(j)
+            }
+        }
         Err(err) => Err(Value::from(err.to_string())),
     }
 }
 
+fn walk(
+    callback: &Value,
+    interpreter: &mut Interpreter,
+    holder: &mut Value,
+    key: Value,
+) -> ResultValue {
+    let mut value = holder.get_field(key.clone());
+
+    if value.is_object() {
+        let obj = value.as_object();
+        if obj.is_some() {
+            let obj = obj.unwrap().clone();
+            for (key, _val) in obj.properties.iter() {
+                let v = walk(callback, interpreter, &mut value, Value::from(key.as_str()));
+                match v {
+                    Ok(v) => {
+                        if !v.is_undefined() {
+                            value.set_field(Value::from(key.as_str()), v);
+                        } else {
+                            value.remove_property(key.as_str());
+                        }
+                    }
+                    Err(_v) => {}
+                }
+            }
+        }
+    }
```

Comment:
```suggestion
    if let Some(obj) = value.clone().as_object() {
        for key in obj.properties.keys() {
            let v = walk(callback, interpreter, &mut value, Value::from(key.as_str()));
            match v {
                Ok(v) if v.is_undefined() => {
                    value.set_field(Value::from(key.as_str()), v);
                }
                Ok(_) => {
                    value.remove_property(key.as_str());
                }
                Err(_) => {}
            }
        }
    }
```
Simplified the `walk` function.

Reply:
Wow! your changes have simplified the monstrosity i wrote above.
However, there are two problems -
1. a clippy warning for clone on line 75 above: `cloned value is neither consumed nor mutated`
2. panic that says, `'GcCell<T> already borrowed'` with this redacted trace:
```
  12: std::panicking::begin_panic_fmt
             at src\libstd/panicking.rs:373
  13: gc::GcCell<T>::borrow_mut
             at boa\src/lib.rs:1
  14: boa::builtins::value::ValueData::set_field
             at boa\src\builtins\value/mod.rs:616
  15: boa::builtins::json::walk
             at boa\src\builtins\json/mod.rs:72
  16: boa::builtins::json::parse
             at boa\src\builtins\json/mod.rs:50
```

Reply:
To be honest, i'd tried using `if let`  with clone combination but because of the above panic, I gave up and was going to ask you how to fix it.

Reply:
I think the issue might be in the difference between cloning first and then getting the object or first getting the object and then cloning. So maybe we could change this:

```rust
    if let Some(obj) = value.clone().as_object() {
        for key in obj.properties.keys() {
```

for this:
```rust
    if let Some(obj) = value.as_object() {
        let obj = obj.clone();
        for key in obj.properties.keys() {
```

Reply:
Unfortunately, `value.as_object()` borrows `value` immutably  and when that borrow is still active, we cant borrow it mutably and pass to walk: `let v = walk(callback, interpreter, &mut value, Value::from(key.as_str()));`

Reply:
```rust
if let Some(obj) = value.as_object().map(|x| x.clone()) {
```
I think this should fix it, right?

Reply:
It did! Learnt a lot of things today! Shout out to you both @Razican and @HalidOdat

Reply:
while refactoring I made a mistake in line 79
```rust
                Ok(v) if v.is_undefined() => {
```

it should be 
```rust
                Ok(v) if !v.is_undefined() => {
```

I have updated the comment :)

Reply:
Clippy has not liked our sneakiness though: `You are needlessly cloning iterator elements`
We also can't use  `cloned()` as clippy suggests [here](https://rust-lang.github.io/rust-clippy/master/index.html#map_clone).

Reply:
> while refactoring I made a mistake at line 79
> 
> ```rust
>                 Ok(v) if v.is_undefined() => {
> ```
> 
> It should be
> 
> ```rust
>                 Ok(v) if !v.is_undefined() => {
> ```
> 
> I updated the comment :)

Oh yeah, i took care of that already.

Reply:
We could use allow attribute to stop clippys complaining!
```rust
#[allow(clippy::map_clone)]
fn walk(
    callback: &Value,
    interpreter: &mut Interpreter,
    holder: &mut Value,
    key: Value,
) -> ResultValue {
```

Reply:
@HalidOdat Brilliant! I'll keep that in mind the next time I wrestle with clippy :)

Reply:
> We could use allow attribute to stop clippys complaining!
> ```rust
> #[allow(clippy::map_clone)]
> fn walk(
>     callback: &Value,
>     interpreter: &mut Interpreter,
>     holder: &mut Value,
>     key: Value,
> ) -> ResultValue {
> ```

Actually, I think it's best if we substitute the `.map(|x| x.clone())` by just `.cloned()`, that should be equivalent, right?

We should try to avoid allowing clippy lints if possible.

Reply:
Sadly `.cloned()` doesn't work it gives this error:
```
error[E0599]: no method named `cloned` found for enum `std::option::Option<gc::GcCellRef<'_, builtins::object::Object>>` in the current scope
   --> boa/src/builtins/json/mod.rs:68:42
    |
68  |       if let Some(obj) = value.as_object().cloned() {
    |                                            ^^^^^^ method not found in `std::option::Option<gc::GcCellRef<'_, builtins::object::Object>>`
    |
    = note: the method `cloned` exists but the following trait bounds were not satisfied:
            `std::option::Option<gc::GcCellRef<'_, builtins::object::Object>>: std::iter::Iterator`
            which is required by `&mut std::option::Option<gc::GcCellRef<'_, builtins::object::Object>>: std::iter::Iterator`

```

Reply:
Found a way to avoid allowing clippy lint:
```rust
    let object = value.as_object().as_deref().cloned();
    if let Some(obj) = object {
```

Reply:
This one nailed it.